### ISSUE TITLE
False positive for Union[type, ...] and inspect.isclass()

### ISSUE THREAD

**Issue by todofixthis** [python/mypy#16640]
# Bug Report
When a variable's type is declared as a union that includes `type` (or `typing.Type[...]`), mypy does not take into account whether usage of that variable occurs inside of an `if inspect.isclass(...)` conditional block, leading to false positives.

## To Reproduce
[gist](https://gist.github.com/mypy-play/ab36fecf249291193a01e84da62a8d61)
[playground](https://mypy-play.net/?mypy=latest&python=3.12&gist=ab36fecf249291193a01e84da62a8d61)

<details>
<summary>Python code</summary>

```python
from inspect import isclass

registry: dict[str, type] = {}

def decorator_factory(arg: str | type):
    if isclass(arg):
        registry[arg.__name__.lower()] = arg
        return arg
    else:
        def decorator(cls: type):
            registry[arg] = cls
            return cls
        return decorator
        
@decorator_factory
class Foo:
    pass
        
@decorator_factory("alt")
class Bar:
    pass
```

</details>

## Expected Behavior**
mypy reports no errors

## Actual Behavior
```
main.py:11: error: Invalid index type "str | type" for "dict[str, type]"; expected type "str"  [index]
Found 1 error in 1 file (checked 1 source file)
```

**Your Environment**
- Mypy version used: 1.7.1
- Mypy command-line flags: (mypy-play defaults)
- Mypy configuration options from `mypy.ini` (and other config files): (mypy-play defaults)
- Python version used: 3.12
